File: src/jvmTest/java/org/snakeyaml/engine/issues/issue17/WindowsTest.java
/*
 * Copyright (c) 2018, SnakeYAML
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations under
 * the License.
 */
package org.snakeyaml.engine.issues.issue17;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static org.junit.jupiter.api.Assertions.fail;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import it.krzeminski.snakeyaml.engine.kmp.api.Load;
import it.krzeminski.snakeyaml.engine.kmp.api.LoadSettings;
import it.krzeminski.snakeyaml.engine.kmp.exceptions.ParserException;
import it.krzeminski.snakeyaml.engine.kmp.scanner.StreamReader;

/**
 * <a href="https://yaml.org/spec/1.2/spec.html#id2774608">https://yaml.org/spec/1.2/spec.html#id2774608</a>
 */
@org.junit.jupiter.api.Tag("fast")
public class WindowsTest {

  Load loader = new Load(LoadSettings.builder().build());

  @Test
  @DisplayName("Check that Windows style line endings handled the same as Unix style ones")
  void testGetLineNumberOnWindows() {
    LoadSettings settings = LoadSettings.builder().build();
    StreamReader reader1 = new StreamReader(settings, "foo\r\nbar");
    StreamReader reader2 = new StreamReader(settings, "foo\nbar");
    reader1.forward(100);
    reader2.forward(100);
    assertEquals(reader1.getLine(), reader2.getLine());
  }

  @Test
  void countLinesCRLF() {
    try {
      loader.loadFromString("\r\n[");
      fail();
    } catch (ParserException e) {
      assertTrue(e.getMessage().contains("line 2,"), e.getMessage());
    }
  }

  @Test
  void countLinesCRCR() {
    try {
      loader.loadFromString("\r\r[");
      fail();
    } catch (ParserException e) {
      assertTrue(e.getMessage().contains("line 3,"));
    }
  }

  @Test
  void countLinesLFLF() {
    try {
      loader.loadFromString("\n\n[");
      fail();
    } catch (ParserException e) {
      assertTrue(e.getMessage().contains("line 3,"));
    }
  }
}

File: src/commonTest/kotlin/it/krzeminski/snakeyaml/engine/kmp/issues/issue17/WindowsTest.kt
package it.krzeminski.snakeyaml.engine.kmp.issues.issue17

import io.kotest.assertions.throwables.shouldThrow
import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.shouldBe
import io.kotest.matchers.string.shouldContain
import it.krzeminski.snakeyaml.engine.kmp.api.Load
import it.krzeminski.snakeyaml.engine.kmp.api.LoadSettings
import it.krzeminski.snakeyaml.engine.kmp.exceptions.ParserException
import it.krzeminski.snakeyaml.engine.kmp.scanner.StreamReader

/**
 * https://yaml.org/spec/1.2/spec.html#id2774608
 */
class WindowsTest : FunSpec({
    val loader = Load()

    test("check that Windows style line endings handled the same as Unix style ones") {
        val settings = LoadSettings()
        val reader1 = StreamReader(settings, "foo\r\nbar")
        val reader2 = StreamReader(settings, "foo\nbar")
        reader1.forward(100)
        reader2.forward(100)
        reader1.line shouldBe reader2.line
    }

    test("count lines CRLF") {
        shouldThrow<ParserException> {
            loader.loadOne("\r\n[")
        }.also {
            it.message shouldContain "line 2,"
        }
    }

    test("count lines CRCR") {
        shouldThrow<ParserException> {
            loader.loadOne("\r\r[")
        }.also {
            it.message shouldContain "line 3,"
        }
    }

    test("count lines LFLF") {
        shouldThrow<ParserException> {
            loader.loadOne("\n\n[")
        }.also {
            it.message shouldContain "line 3,"
        }
    }
})
